[BUGFIX] Selector precedence pseudos/attributes

JakeQZ opened the pull request:
Corrected selector precedence calculation for selectors involving
pseudo-classes and/or attributes (ref:
https://developer.mozilla.org/en-US/docs/Web/CSS/Specificity).

Increased multiplier for selector type to allow up to 100 of each type.

Added PHPUnit tests for selector precedence.  Moved common testing HTML to a
constant.

oliverklee commented:
Hi @JakeQZ, thanks for the PR! :heart: 

Would you be willing to create another PR for the testing HTMLconstant? Then we can have the cleanup nicely separated from the actual bugfix, which will help keep the Git history (and line-based history) clean and would allow for separate reverts (should we every need to revert any of those commits)?

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -672,19 +676,93 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+            'id more specific than class and types' => ['#p4', 'html body .p-4', '>', '<p class="p-4" id="p4"'],
+            'id more specific than 20 classes and types'
+                => ['#p4', 'html body ' . str_repeat('.p-4', 20), '>', '<p class="p-4" id="p4"'],
+            'psuedo-class as specific as class' => ['.p-1', '*:first-child', '==', '<p class="p-1"'],
```
typo: psuedo -> pseudo

JakeQZ replied:
Have corrected typo.

JakeQZ commented:
Ok, have created new PR for the testing HTML constant.

JakeQZ commented:
Have corrected typo, and merged the now separate HTML constant cleanup and changed the opening comment of this PR to reflect that it doesn't include that.

oliverklee commented on CHANGELOG.md:
```diff
@@ -24,6 +24,8 @@ This project adheres to [Semantic Versioning](https://semver.org/).
 
 ### Fixed
+- Calculation of selector precedence for selectors involving pseudo-classes
+  and/or attributes.
```
Could you please add a link to the PR (so this is consistent with the other changelog entries)? Then we don't need to edit your PR before we merge it. Thanks!

JakeQZ replied:
Done

oliverklee commented on src/Emogrifier.php:
```diff
@@ -1404,18 +1404,24 @@ private function getCssSelectorPrecedence($selector)
+                '(?:\\.|\\[|(?<!:):(?!not\\())',
+                // elements (not attribute values or `:not`), pseudo-elements: worth 1
+                '(?:(?<![="\':\\w-])|::)'
+            ];
```
I propose putting the values (1, 100, 10000) in code, e.g. as array keys. Then this array does not depend on the order of the entries.

JakeQZ replied:
I've put the values as array values, so that it would be possible in future to add additional regular expressions for the same value if desired.  There is still some dependency on the order of the entries, as the matcher for element type relies on the ID and class matches having been removed.

In fact, it could still match after a space in an attribute value expression, such as `[title="buenas dias"]` - maybe this should also be addressed while we're here...?

oliverklee commented on src/Emogrifier.php:
```diff
@@ -1404,18 +1404,24 @@ private function getCssSelectorPrecedence($selector)
                 }
                 $number = 0;
-                $selector = preg_replace('/' . $s . '\\w+/', '', $selector, -1, $number);
+                $selector = preg_replace('/' . $s . '[\\w-]++/', '', $selector, -1, $number);
```
Is this related to this PR? And if it is not, could you please undo this change?

And if this actually fixes a but, I'd prefer this change to have its own PR (including changelog entry and unit tests).

JakeQZ replied:
I've reverted this change, as none of the tests I added fail without it.  When removing matched parts, I wanted to see `:first-child` rather than just `:first` removed, but maybe that isn't necessary.  I'll see if I can come up with a valid test case that would require it...

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
     }
 
+    /**
+     * Data provider for {@see EmogrifierTest::emogrifyAppliesMostSpecificSelectorCssToMatchingElements}
```
I don't think we need the description and the `@see` here as the method name is "speaking" enough.

JakeQZ replied:
Removed

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+    /**
+     * Data provider for {@see EmogrifierTest::emogrifyAppliesMostSpecificSelectorCssToMatchingElements}
+     *
+     * @return string[][]
```
Good type!

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+            $parameterSets[$name . ' (2nd selector in last rule)'] = [
+                $firstRuleBlock . ' ' . $secondRuleBlock,
+                $data[2] === '>' ? $firstRuleBlockWasApplied : $secondRuleBlockWasApplied
+            ];
```
Instead of having code blocks with comments explaining what they do, please move these code blocks to private methods with "speaking" names. This will make it easier to get an overview of what happens here.

JakeQZ replied:
Refactored into new private methods `getMatchedCssDataSet` and `getSelectorSpecificityCssDataSet`

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+                => ['*:not(.p-2)', 'html body p', '>', '<p class="p-1"'],
+            ':not with class more specific than :not with types'
+                => ['*:not(.p-2)', '*:not(html body h1)', '>', '<p class="p-1"'],
+            'class more specific than :not with types' => ['.p-1', '*:not(html body h1)', '>', '<p class="p-1"'],
```
Great test case descriptions!

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -684,6 +684,92 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+         */
+        $parameterSets = [];
+        foreach ($selectorData as $name => $data) {
+            $firstRuleBlock = $data[0] . ' { %1$s }';
```
Maybe also move the code in the loop into a well-named method.

JakeQZ replied:
Refactored loop body into new private method `getSelectorSpecificityCssDataSets`

oliverklee commented:
(Just removed some comments that don't apply anymore so we keep this PR tidy.)

JakeQZ commented:
I'm updating the branch with the requested changes, but will also need to rebase because of CHANGELOG.md.  I've responded to some of the individual comments with more detail.

I've also spotted that, I think, a space in an attribute value e.g. `[title="buenas dias"]` will still result in an additional match as an element type (`dias`) - maybe this should also be addressed while we're here...?

JakeQZ commented:
Have rebased and merged, and confirmed that `[title="buenas dias"]` is being incorrectly evaluated as more specific becuase of the space...

oliverklee commented:
> I've also spotted that, I think, a space in an attribute value e.g. [title="buenas dias"] will still result in an additional match as an element type (dias) - maybe this should also be addressed while we're here...?

Please let's fix this in a separate PR.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+     * @param string $firstSelector A selector expression to be used in an earlier CSS declaration
+     * @param string $secondSelector Some other selector expression to be used in a later CSS declaration
+     * @param bool $firstSelectorIsMoreSpecific Whether the first selector's rules are expected to prevail
+     * @param string $matchedTagPart Tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by both
```
I don't think we need the `@see` here.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
 
+    /**
+     * @param string $selector A selector expression
+     * @param string $matchedTagPart Tag (in {@see EmogrifierTest::COMMON_TEST_HTML}) expected to be matched by the
```
I don't think we need the `@see` here.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+    private function getSelectorSpecificityCssDataSet(
+        $firstSelector,
+        $secondSelector,
+        $firstSelectorIsMoreSpecific,
```
Instead of having this "toggle" parameter (that toggles between two different behaviors), we should have two separate, well-named methods.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+    {
+        $selectorData = [
+            ['class more specific than types', '.p-1', 'html body p', '>', '<p class="p-1"'],
+            ['id more specific than class and types', '#p4', 'html body .p-4', '>', '<p class="p-4" id="p4"'],
```
id -> ID (in human-readable text, not in code)

JakeQZ replied:
Changed

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+
+        $datasets = [];
+        foreach ($selectorData as $data) {
+            $datasets += call_user_func_array([$this, 'getSelectorSpecificityCssDataSets'], $data);
```
Why not call $this->getSelectorSpecificityCssDataSets directly?

JakeQZ replied:
No longer applicable after refactoring

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+        ];
+
+        $datasets = [];
+        foreach ($selectorData as $data) {
```
I think we need different variable names here because currently both "data" and "selectorData" basically mean the same thing, but the names should make the difference clear.

JakeQZ replied:
No longer applicable after refactoring

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,143 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+                $firstSelector,
+                $secondSelector,
+                $operator === '>',
+                $matchedTagPart
```
We can simplify this by having 2 or 3 separate tests (with shared or separate data providers): one for selectors that have the same specificity (what a word!), and one or two for different specificities (depending on whether you think that we order matters).

What do you think?

JakeQZ replied:
Have refactored along those lines.

oliverklee commented on src/Emogrifier.php:
```diff
@@ -1404,18 +1404,26 @@ private function getCssSelectorPrecedence($selector)
+             *                    values are a multiplier factor per match to weight specificity
+             */
+            $search = [
+                // ids: worth 10000
```
ids -> IDs

JakeQZ replied:
Changed

JakeQZ commented:
> We can simplify this by having 2 or 3 separate tests (with shared or separate data providers): one for selectors that have the same specificity (what a word!), and one or two for different specificities (depending on whether you think that we order matters).
> 
> What do you think?

I think doing that would also resolve the other issues with the testing code you commented on.  So I will refactor with that approach...

JakeQZ commented:
I've refactored as suggested, and changed "id" to "ID" in comments/dataset names.  This has also reduced the amount of code a bit, and the number of distinct tests (pairs of tests now combined, and checking the validity of the dataset parameters was only necessary in 1/4 cases).

oliverklee commented on src/Emogrifier.php:
```diff
@@ -1404,18 +1404,26 @@ private function getCssSelectorPrecedence($selector)
+             * @var int[] $search Keys are a regular expression part to match before a CSS name,
+             *                    values are a multiplier factor per match to weight specificity
+             */
+            $search = [
```
This now even can be a constant instead of a local variable.

oliverklee replied:
(which would make the method more concise and move fixed data out of it)

JakeQZ replied:
Have moved it to a constant.

oliverklee commented on src/Emogrifier.php:
```diff
@@ -1404,18 +1420,13 @@ private function getCssSelectorPrecedence($selector)
-            $search = ['\\#', '\\.', ''];
-
-            foreach ($search as $s) {
+            foreach ($this->selectorPrecedenceMatchers as $s => $value) {
```
While we're touching the foreach anyway, please let's rename `$s` to something more expressive, e.g. `$matcher`.

JakeQZ replied:
Done

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,118 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+        return [
+            'class more specific than types' => ['html body p', '.p-1', '<p class="p-1"'],
+            'ID more specific than class and types' => ['html body .p-4', '#p4', '<p class="p-4" id="p4"'],
+            'ID more specific than 20 classes and types' => [
```
Why exactly 20? Could this also be a different number? Maybe a comment would be helpful why this number is here.

JakeQZ replied:
20 was arbitrary.  Have bumped this to 99 which is the maximum supported (without rewriting the comparator, and I don't think likely to be a real-world limitation).  Also added a comment suggesting a similar test for classes vs 99 types, which would require support for chaining pseudo-classes, e.g. `:not(h1):not(h1)...`

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,118 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+     */
+    public function emogrifyWouldApplyLessSpecificCssSelectorToMatchingElements( // phpcs:ignore
+        $lessSpecificSelector,
+        // PHP_CodeSniffer warning ignored: The method parameter $moreSpecificSelector is never used
```
You can switch the order of parameters and then make this test method have one argument less than the others (having surplus parameters in a method call is fine in PHP).

I propose this order:
`$matchedTagPart, $lessSpecificSelector, $moreSpecificSelector`

JakeQZ replied:
Good thinking (it took some coaxing to get CodeSniffer to suppress the warning without introducing a different one!).  Agree with your proposed order as being most logical and have changed this and the other methods for consistency.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -690,6 +690,118 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+            $lessSpecificSelector . ' { color: red } ' .
+            $moreSpecificSelector . ' { color: green } ' .
+            $moreSpecificSelector . ' { background-color: green } ' .
+            $lessSpecificSelector . ' { background-color: red }'
```
Any particular reason why you're leaving out the closing semicolon (`color: red;`)?

JakeQZ replied:
No particular reason.  Have added it.

JakeQZ commented:
Have made the suggested changes and rebased.  More detail in reply to individual comments.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -737,6 +737,117 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+    {
+        return [
+            'class more specific than types' => ['<p class="p-1"', 'html body p', '.p-1'],
+            'ID more specific than class and types' => ['<p class="p-4" id="p4"', 'html body .p-4', '#p4'],
```
Do we need two types here, or would one suffice? (Test cases should be minimal so only relevant things are there, making it easier to find out what's important when reading the test.)

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -737,6 +737,117 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+    {
+        return [
+            'pseudo-class as specific as class' => ['<p class="p-1"', '.p-1', '*:first-child'],
+            'attribute as specific as class' => ['<span title="bonjour"', 'span[title="bonjour"]', '.p-2 span'],
```
Should we also have a case comparing a pseudo-class comparing an attribute, or would that be redundant?

oliverklee commented:
> I'll go through them with a finer-toothed comb - probably won't be until early next week now...

Okay, I'll do the review then.

Could you also please rebase?

JakeQZ commented:
I've reworked the test data as planned, and rebased.

JakeQZ commented:
> I've reworked the test data as planned, and rebased.

Actually I think the test data for selector specificity inequality/ordering can be simplified - if all selectors can target the same element, then it can be generated from an array of selectors of increasing specificity, avoiding repetition.

Please hold off reviewing this for now, and I'll either commit another change, or advise that it can't be simplified in this way.

JakeQZ commented:
> Actually I think the test data for selector specificity inequality/ordering can be simplified - if all selectors can target the same element, then it can be generated from an array of selectors of increasing specificity, avoiding repetition.

I have now changed this as described; the presentation of the test data in the code should now be clearer.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -737,6 +737,198 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+    {
+        // broken: testing selectors with 99 types or 99 IDs requires support for chaining `:not(h1):not(h1)...` or
+        // `#p4#p4...`, so currently 2 types and/or 2 IDs are tested as 'many' instead of 99 of each.
+        $classP4Repeated = str_repeat('.p-4', 99);
```
By the way, please let's reduce the 99 to the minimal number that's relevant for the effect. My educated guess is that it's 2 or 3 (as this is not a performance stress test).

JakeQZ replied:
I've changed this from 99 to 2 in these tests, as testing that 'many' is more specific than one is sufficient.

To confirm the increase in selector type weighting I've added a single separate dataset with 99.

JakeQZ commented:
I've changed 99 to 2 in the selector inequality test data for confirming 'many' is more specific than 'one', and added a single separate test with 99 classes to confirm the increase in selector weighting multiplier.

oliverklee commented on tests/Unit/EmogrifierTest.php:
```diff
@@ -737,6 +737,204 @@ public function emogrifyNotAppliesCssToNonMatchingElements($css, $expectedHtml)
+     * @param string $lessSpecificSelector A selector expression
+     * @dataProvider differentCssSelectorSpecificityDataProvider
+     */
+    public function emogrifyWouldApplyLessSpecificCssSelectorToMatchingElements(
```
The way I read this test is that is doesn't really check for less/more specific selectors, but tests that the given selectors work (without regard to specificity), i.e., we can remove this test as we already have tests for the individual selectors matching. Or am I missing anything here?

JakeQZ replied:
Yes, this was just to confirm that the test datasets did in fact represent valid tests, and the lower precedence selector provided would in fact match the intended target if used alone.  However, since the test data is now written such that the lower precedence selector will be the same as the higher precedence one from the previous dataset, this is now unnecessary apart from the first dataset - but that uses the universal selector which obviously matches - and the final separate dataset with 99 classes - though for just a single dataset this could be confirmed 'by inspection'.

So I have removed this.
